stubs/docker/docker/utils/utils.pyi

```diff
@@ -1,34 +1,67 @@
-from _typeshed import Incomplete
-from typing import NamedTuple
+import datetime
+from _typeshed import FileDescriptorOrPath, Incomplete, ReadableBuffer
+from collections.abc import Iterable, Mapping
+from typing import Literal, NamedTuple, NoReturn, Protocol, TypeVar, overload, type_check_only
+
+from ..tls import TLSConfig
+
+_T = TypeVar("_T")
+_K = TypeVar("_K")
+_V = TypeVar("_V")
+
+# This Protocol is taken from stdlib/shlex.pyi
+@type_check_only
+class _ShlexInstream(Protocol):
+    def read(self, size: Literal[1], /) -> str: ...
+    def readline(self) -> object: ...
+    def close(self) -> object: ...
 
 class URLComponents(NamedTuple):
-    scheme: Incomplete
-    netloc: Incomplete
-    url: Incomplete
-    params: Incomplete
-    query: Incomplete
-    fragment: Incomplete
+    scheme: str | None
+    netloc: str | None
+    url: str
+    params: str | None
+    query: str | None
+    fragment: str | None
 
-def create_ipam_pool(*args, **kwargs) -> None: ...
-def create_ipam_config(*args, **kwargs) -> None: ...
-def decode_json_header(header): ...
-def compare_version(v1, v2): ...
-def version_lt(v1, v2): ...
-def version_gte(v1, v2): ...
-def convert_port_bindings(port_bindings): ...
-def convert_volume_binds(binds): ...
-def convert_tmpfs_mounts(tmpfs): ...
-def convert_service_networks(networks): ...
-def parse_repository_tag(repo_name): ...
-def parse_host(addr, is_win32: bool = False, tls: bool = False): ...
-def parse_devices(devices): ...
-def kwargs_from_env(environment: Incomplete | None = None): ...
-def convert_filters(filters): ...
-def datetime_to_timestamp(dt): ...
-def parse_bytes(s): ...
+def create_ipam_pool(*args, **kwargs) -> NoReturn: ...
+def create_ipam_config(*args, **kwargs) -> NoReturn: ...
+def decode_json_header(header: str | ReadableBuffer): ...
+def compare_version(v1: str, v2: str) -> Literal[0, -1, 1]: ...
+def version_lt(v1: str, v2: str) -> bool: ...
+def version_gte(v1: str, v2: str) -> bool: ...
+def convert_port_bindings(
+    port_bindings: dict[object, Incomplete],  # we are converts key to a string using str()
+) -> dict[str, list[dict[str, str]]]: ...
+@overload
+def convert_volume_binds(binds: list[_T]) -> list[_T]: ...
+@overload
+def convert_volume_binds(binds: dict[str | bytes, Incomplete]) -> list[str]: ...
+@overload
+def convert_tmpfs_mounts(tmpfs: dict[_K, _V]) -> dict[_K, _V]: ...
+@overload
+def convert_tmpfs_mounts(tmpfs: list[str]) -> dict[str, str]: ...
+@overload
+def convert_service_networks(networks: None) -> None: ...
+@overload
+def convert_service_networks(networks: list[str | _T]) -> list[_T]: ...
```

Comment:
I don't think this is correct, a `list[str]` should result in a `list[dict[str, str]]`

Also `list` is invariant, so you have to expand the union to allow all possible permutations.

Either:
```suggestion
def convert_service_networks(networks: list[str] | list[dict[str, str]] | list[str | dict[str, str]]) -> list[dict[str, str]]: ...
```
Or at least binding a TypeVar to `dict[str, str]`.